### PR TITLE
`Exam mode`: Show a validation message for a missing exam title

### PR DESCRIPTION
**Pull request by HannesHarbeck**
### Summary
The exam create/edit form now flags an empty title with a red "missing/invalid" message like the date fields, and the server rejects a null or blank exam title with a 400 on create, update and import. This closes two server gaps: an empty-string title was persisted, and a crafted null title threw a 500 while being mapped to the entity.

#### General
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.tum.de/developer/guidelines/language).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.tum.de/developer/development-process#pr-naming-conventions).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.tum.de/developer/guidelines/performance) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.tum.de/developer/guidelines/server-development) and the [REST API guidelines](https://docs.artemis.tum.de/developer/guidelines/rest-api).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding guidelines](https://docs.artemis.tum.de/developer/guidelines/client-development).
- [x] I **strictly** followed the [AET UI-UX guidelines](https://ls1intum.github.io/ui-ux-guidelines/).
- [x] Following the [theming guidelines](https://docs.artemis.tum.de/developer/guidelines/client-theming), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Vitest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.tum.de/developer/guidelines/client-tests).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
The exam create and edit form only disabled the save button when the title was empty, without the missing/invalid hint shown for the required date fields. A user emptying the title got no visible feedback about why saving was blocked.

On top of that, the server did not enforce the requirement it relies on: the `title` column is `nullable = false`, which only catches `null`, so an empty-string title was persisted, and a crafted request with a `null` title threw a 500 (NPE) while mapping the DTO to the entity.

### Description
- **Client:** the exam form now renders a red "'Title' is missing/invalid" message under the title field while the title is empty, mirroring the date fields' behaviour and reusing the same `text-state-danger` styling.
- **Server:** `ExamResource` rejects a `null` or blank title with `400 examTitleEmpty` on create, update and import, validated on the request before it is mapped to an entity (so the `null` case can no longer NPE in `Exam#setTitle`).

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Course

1. Log in as the instructor and open Exam Management for the course.
2. Click "Create a new exam".
3. Leave the **Title** empty. A red "'Title' is missing/invalid" message appears under the title field, just like the date fields (German: "'Titel' fehlt/ungültig").
4. Type a title. The message disappears, and once the required dates are valid, saving works.
5. Delete the title again. The message reappears and the save button is disabled.

Note: this PR only changes the instructor-facing exam configuration form and the server-side validation. The student exam-taking UI (exam mode) is not affected.

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage

**Note:** Some tests in the [Test job](https://github.com/ls1intum/Artemis/actions/runs/33312810035) did not pass (`failure`). Coverage below may be partial.

#### Client

| Class/File | Line Coverage | Lines | Expects | Ratio |
|------------|-------------:|------:|--------:|------:|
| exam-update.component.ts | 92.26% | 430 | 203 | 47.2 |

#### Server

| Class/File | Line Coverage | Lines |
|------------|-------------:|------:|
| ExamResource.java | not found (modified) | 877 |

_Last updated: 2026-08-30 14:02:55 UTC_

### Screenshots
_To be added: the exam create form with an empty title showing the red "'Title' is missing/invalid" message (light and dark theme)._

## Summary by CodeRabbit

* **Bug Fixes**
  * Exam creation, updates, and imports now reject missing or blank titles with a clear validation error.
  * Added inline feedback when an exam title is missing or invalid.
  * Saving is disabled until the exam title contains valid text.
  * Added English and German translations for the title validation message.

* **Tests**
  * Added coverage for title validation in the exam interface and backend, including missing and whitespace-only titles.

image

image

image

### CODE REVIEW

**Review by Claudia-Anthropica — CHANGES_REQUESTED** [review]
@HannesHarbeck The title validation is consistently applied to create, update, and import and has focused client and server coverage. However, the required `All required CI Passed` check is failing because `E2E / Report E2E Overall Status` failed; please investigate the E2E failure and restore the required check before approval.

File: src/main/webapp/app/exam/manage/exams/update/exam-update.component.html [hunk 1]
```diff
@@ -25,6 +25,9 @@ <h5 class="pb-1" jhiTranslate="artemisApp.examManagement.sections.configuration"
                     [hideChannelName]="hideChannelNameInput()"
                     [initChannelName]="isImport() || exam.id === undefined"
                 />
+                @if (!exam.title?.trim()) {
```
Comment (Claudia-Anthropica) [inline src/main/webapp/app/exam/manage/exams/update/exam-update.component.html:28]: @HannesHarbeck [medium] A whitespace-only title reaches this branch, but Angular's `required` validator considers every non-empty string valid, while `isValidConfiguration` does not check the title. Consequently, Save can remain enabled while this invalid message is visible; include the trimmed-title check in the save validity/disabled logic and add a test asserting that a whitespace-only title disables Save.

**Review by Claudia-Anthropica — CHANGES_REQUESTED** [review]
@HannesHarbeck Required CI now passes, and the server validation covers create, update, and import. One client-side gap remains: a whitespace-only title displays the invalid message but leaves Save enabled, resulting in a rejected server request instead of preventing submission.

**Review by Claudia-Anthropica — DISMISSED** [review]
@HannesHarbeck The current head rejects null, empty, and whitespace-only titles across create, update, and import, with focused server tests asserting the `examTitleEmpty` contract. The client now applies the same trimmed-title rule to the validation message and Save-button validity, including regression coverage for the previously reported whitespace-only case. The snapshot contains no required checks and no completed failing checks.

File: src/main/webapp/app/exam/manage/exams/update/exam-update.component.html (hunk 1 above)
Comment (Claudia-Anthropica) [reply src/main/webapp/app/exam/manage/exams/update/exam-update.component.html:28]: @HannesHarbeck Thanks, this is fixed now: the trimmed-title check participates in `isValidConfiguration`, and the regression test verifies that Save is disabled for whitespace-only titles.

**Review by SedaOran — DISMISSED** [review]
Code LGTM. Approved. Backend rejects null/blank/whitespace titles on create/update/import with a clean 400 (examTitleEmpty); frontend trim() check gates the message + Save button. Tests cover all three endpoints, and I verified the UI locally (blank → message + Save disabled)

**Comment by SultanTheHL** [comment]
Tested locally and works as described. 
image

However, I think it would look better if the error message was closer to the input field(like with the dates)

**Review by SultanTheHL — CHANGES_REQUESTED** [review]
comment above

**Review by Claudia-Anthropica — DISMISSED** [review]
@HannesHarbeck The server consistently rejects null, empty, and whitespace-only exam titles across create, update, and import, with focused contract tests. The client applies the same trimmed-title rule to its message and Save-button state, and the latest layout adjustment uses an available spacing utility. The snapshot lists no required checks and no completed failing checks.

**Comment by HannesHarbeck** [comment]
@SultanTheHL I adjusted the layout accordingly

**Review by SedaOran — DISMISSED** [review]
Now UI works well and It's aligned. Approving.  
Image

File: src/main/java/de/tum/cit/aet/artemis/exam/web/ExamResource.java [hunk 2]
```diff
@@ -557,6 +560,18 @@ private void checkExamTitleLengthElseThrow(Exam exam) {
...
 
+    /**
+     * Checks that the exam title is present, so an exam is never created or updated with a missing or blank title. The client marks this too, but crafted requests and import
+     * payloads bypass the UI. This validates the raw request title before it is mapped to an entity, because {@link Exam#setTitle} would throw on a null title during mapping.
+     *
+     * @param title the exam title from the request
+     */
+    private void checkExamTitleIsPresentElseThrow(String title) {
+        if (title == null || title.isBlank()) {
+            throw new BadRequestAlertException("The exam title must not be empty.", ENTITY_NAME, "examTitleEmpty");
```
Comment (WoH) [inline src/main/java/de/tum/cit/aet/artemis/exam/web/ExamResource.java:572]: examTitleEmpty has no error.json entry in en or de. 

checkExamTitleLengthElseThrow right above this (examTitleTooLong) has both, and so does every other exam validation error in this class (noCourse, wrongCourseId, examTimes, negativePoints, correctionRoundViolation, attendanceCheckViolation).

confirmed at runtime: fetch('/i18n/en.json') and de.json both lack error.examTitleEmpty while error.examTitleTooLong is present in both. means a crafted request or any non-UI caller gets the raw english exception title in the toast, untranslated for german users. two-line fix, same pattern as examTitleTooLong.

**Review by WoH — DISMISSED** [review]
tested commit 2b13db6. setup: local dev, e2e seed, instructor artemis_test_user_16, course 9023 E2E Exam Management Course. 


1. create exam, leave title empty, red 'Title' is missing/invalid message under the field like the date fields: yes
pr13362-03-empty-title-message


2. type a title, message disappears, save enabled once dates valid too: yes, per-keystroke, no fill()
pr13362-06-valid-title-save-enabled


3. delete the title again, message reappears, save disabled: yes, tested on both create and the edit form (step not in your list but same code path)

pr13362-08-edit-save-disabled
pr13362-07-edit-title-cleared


[pr13362-v01-create-exam-title-validation.webm](https://github.com/user-attachments/assets/335665f3-56e8-41ed-9e94-7c3503074871)

[pr13362-v02-edit-exam-clear-title.webm](https://github.com/user-attachments/assets/8f05e5f3-c1f4-4653-9cc2-5f42fb9aad45)

also checked:
- whitespace-only title (3 spaces, real keystrokes + blur): native `required` reports the field valid (green border) but your trim check still shows the message and blocks save. isolated it by filling all three dates first so title was the only failing field. this is the actual bug the PR fixes, good catch.

pr13362-04-whitespace-title
pr13362-05-dates-filled-title-blank

- full create round trip: POST /api/exam/courses/9023/exams -> 201, title matches in request/response/UI, no console errors, no toast

- crafted PUT with title:null on the saved exam -> 400, error.examTitleEmpty, exam title unchanged after (confirmed by reload). see inline comment, this error key has no en/de translation, unlike its neighbors

**Review by Claudia-Anthropica — DISMISSED** [review]
@HannesHarbeck The exact head consistently rejects null/blank titles for create, update, and import, while the client blocks whitespace-only titles with focused coverage. The failed E2E test concerns post-reload quiz-answer re-sending; its only call through the changed create endpoint uses a valid title, so this change does not reach the failed behavior. `Test / Server Tests (PostgreSQL)` and therefore `All required CI Passed` remain red, but the captured server evidence names no failing test or compiler error, so that failure cannot be attributed. The existing unresolved thread identifies one [low] localization gap: `error.examTitleEmpty` is absent from the English and German `error.json`, leaving the raw English fallback for non-UI invalid requests.

**Review by ShudongCai — DISMISSED** [review]
Tested locally, works as expected

Image

Image

**Review by SedaOran — DISMISSED** [review]
Tested locally, works as expected. Approved.

**Review by LukaPetrovicTUM — DISMISSED** [review]
Tested locally. Works as expected. Code looks good.

**Comment by LukaPetrovicTUM** [comment]
Warning under the title input:

01_input_warning

Disabled button:

02_diabled_button

Warning disappears when title is entered:

03_valid_title

File: src/main/java/de/tum/cit/aet/artemis/exam/web/ExamResource.java (hunk 2 above)
Comment (HannesHarbeck) [reply src/main/java/de/tum/cit/aet/artemis/exam/web/ExamResource.java:572]: Good catch, fixed in f395d381f3: added examTitleEmpty to both en/error.json and de/error.json (mirroring examTitleTooLong), so the message is now localized instead of falling back to the raw English exception title.
